scripts/check_compliance.py

```diff
@@ -211,6 +215,124 @@ def run(self):
 """.format(max_top_items, n_top_items, __file__)
 
 
+        #
+        # Look for references to undefined Kconfig symbols outside Kconfig
+        # files
+        #
+
+        # Grep for symbol references.
+        #
+        # Example output line for a reference to CONFIG_BAZ at line 17 of
+        # foo/bar.c:
+        #
+        #   foo/bar.c<null>17<null>CONFIG_BAZ
+        #
+        # Skip the samples/ and tests/ directories for now. They often contain
+        # Kconfig files that are not part of the main Kconfig tree, which will
+        # trigger false positives until we do something fancier. Skip
+        # doc/releases too, which often references removed symbols.
+        grep_cmd = "git grep --only-matching --line-number -I --null " \
+                   "--extended-regexp --word-regexp CONFIG_[A-Z0-9_]+ " \
+                   "-- :!samples :!tests :!doc/releases"
+
+        grep_process = subprocess.Popen(grep_cmd.split(),
+                                        stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE,
+                                        cwd=self.zephyr_base)
+
+        grep_stdout, grep_stderr = grep_process.communicate()
+        if grep_process.returncode:
+            self.case.result = Error("error running 'git grep'", "error")
+            # grep_stderr will be 'bytes' here, which prints as b'oops'
+            self.case.result._elem.text = \
+                "'{}' failed with exit code {} and wrote {} to stderr " \
+                "(while searching for Kconfig symbol references)" \
+                .format(grep_cmd, grep_process.returncode, grep_stderr)
+            return
+
+        defined_syms = set(sym.name for sym in kconf.unique_defined_syms)
+        undef_to_locs = collections.defaultdict(list)
+
+        # splitlines() supports various line terminators
+        for line in grep_stdout.decode("utf-8").splitlines():
+            path, lineno, sym_name = line.split("\0")
+
+            # [7:] removes the "CONFIG_" prefix
+            if sym_name[7:] not in defined_syms and \
+               sym_name not in UNDEF_KCONFIG_WHITELIST:
+
+               undef_to_locs[sym_name].append("{}:{}".format(path, lineno))
+
+        if undef_to_locs:
+            undef_desc = "\n".join(
+                "{:35} {}".format(sym_name, ", ".join(locs))
+                for sym_name, locs in sorted(undef_to_locs.items()))
+
+            self.case.result = Failure("references to undefined Kconfig symbols",
+                                       "failure")
+            self.case.result._elem.text = \
+                "Found references to undefined Kconfig symbols. If any of " \
+                "these are false positives, then add them to " \
+                "UNDEF_KCONFIG_WHITELIST in {} in the ci-tools repo.\n\n{}" \
+                .format(__file__, undef_desc)
```

Comment:
Each independent section has a

```
#
# Blah blah
#
```

I was thinking of splitting those out into separate functions, but didn't get around to it yet. You'd want to reuse at least the `kconf` object to avoid having to reparse stuff, but could pass it as a parameter.

Reply:
Hmm... also just remembered that the stuff in https://github.com/zephyrproject-rtos/ci-tools/pull/26 will make it less awkward, so I'll wait for that.

Reply:
Rebased now, and put into a separate function.